network/rcmgr.go

```diff
@@ -0,0 +1,317 @@
+package network
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+// ResourceManager is the interface to the network resource management subsystem.
+// The ResourceManager tracks and accounts for resource usage in the stack, from the internals
+// to the application, and provides a mechanism to limit resource usage according to a user
+// configurable policy.
+//
+// Resource Management through the ResourceManager is based on the concept of Resource
+// Management Scopes, whereby resource usage is constrained by a DAG of scopes,
+// The following diagram illustrates the structure of the resource constraint DAG:
+// System
+//   +------------> Transient.............+................+
+//   |                                    .                .
+//   +------------>  Service------------- . ----------+    .
+//   |                                    .           |    .
+//   +------------->  Protocol----------- . ----------+    .
+//   |                                    .           |    .
+//   +-------------->  Peer               \           |    .
+//                      +------------> Connection     |    .
+//                      |                             \    \
+//                      +--------------------------->  Stream
+//
+// The basic resources accounted by the ResourceManager include memory, streams, connections,
+// and file  descriptors. These account for both space and time used by
+// the stack, as each resource has a direct effect on the system
+// availability and performance.
+//
+// The modus operandi of the resource manager is to restrict resource usage at the time of
+// reservation. When a component of the stack needs to use a resource, it reserves it in the
+// appropriate scope. The resource manager gates the reservation against the scope applicable
+// limits; if the limit is exceeded, then an error (wrapping ErrResourceLimitExceeded) and it
+// is up the component to act accordingly. At the lower levels of the stack, this will normally
+// signal a filure of some sorts, like failing to opening a stream or a connection, which will
```

Comment:
Did you consider defining a normalized errors for components to wrap when the resource manager denies an action, and this prevents progress?

Reply:
Yes, we have ErrResourceLimitExceeded which implements net.Error as temporary.

I should probably add something here about it.

Reply:
wait, there is a reference to ErrResourceLimitExceeded already here; it is defined in errors.go